### PR TITLE
[WIP] AJD-238 scenario validation

### PR DESCRIPTION
**Pull request by WJaworskiRobotec**
## Types of PR

- [X] New Features
- [ ] Upgrade of existing features
- [ ] Bugfix

## Description
Added sample validation of scenario files

### CODE REVIEW

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 1]
```diff
@@ -18,10 +18,54 @@
...
+def find_in_dict(key, dictionary):
+    values = list()
+    if key in dictionary:
+        values.append(dictionary[key])
+    for k, v in dictionary.items():
+        if k == key:
+            values.append(v)
+        if isinstance(v, list):
+            for i in v:
+                for j in find_in_dict(key, i):
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:35]: How do you know that `i` is a dictionary?

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 2]
```diff
@@ -18,10 +18,54 @@
...
+                    values.append(j)
+        elif isinstance(v,dict):
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:37]: `isinstance(v,dict)` -> `isinstance(v, dict)` (space)

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 3]
```diff
@@ -18,10 +18,54 @@
...
+        return None
+
+    def get_warning_message(self):
+        return self.warning_message
+
+
+class ReachPositionConditionValidator(ScenarioValidator):
+    def __init__(self):
+        self.warning_message = "ReachPositionCondition position does not match the last position of the path"
+        self.schema = xmlschema.XMLSchema(
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:56]: Shouldn't it be a member of a base class? I imagine that every scenario validator will need this schema

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 4]
```diff
@@ -18,10 +18,54 @@
...
+            resource_string(__name__, "resources/OpenSCENARIO.xsd").decode("utf-8")
+        )
+
+    def __call__(self, path: Path):
+        scenario = self.schema.to_dict(str(path))
+        #Search for the last Routing Action destionation:
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:62]: `#Search` -> `# Search`(space)

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 5]
```diff
@@ -18,10 +18,54 @@
...
+        last_destination = find_in_dict('Position',find_in_dict('RoutingAction', scenario)[-1])
+
+        #Search for ReachPositionCondition:
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:65]: `#Search` -> `# Search`(space)

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 6]
```diff
@@ -18,10 +18,54 @@
...
+    def __init__(self):
+        self.warning_message = "ReachPositionCondition position does not match the last position of the path"
+        self.schema = xmlschema.XMLSchema(
+            resource_string(__name__, "resources/OpenSCENARIO.xsd").decode("utf-8")
+        )
+
+    def __call__(self, path: Path):
+        scenario = self.schema.to_dict(str(path))
+        #Search for the last Routing Action destionation:
+        last_destination = find_in_dict('Position',find_in_dict('RoutingAction', scenario)[-1])
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:63]: `'Position',find_in_dict` -> `'Position', find_in_dict`(space)

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 7]
```diff
@@ -18,10 +18,54 @@
...
+
+        #Search for ReachPositionCondition:
+        reach_position_conditon = find_in_dict('Position',find_in_dict('ReachPositionCondition', scenario)[-1])
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:66]: `'Position',find_in_dict` -> `'Position', find_in_dict` (space)

File: openscenario/openscenario_utility/openscenario_utility/validation.py [hunk 8]
```diff
@@ -18,10 +18,54 @@
...
+
+        return reach_position_conditon == last_destination
```
Comment (danielm1405) [inline openscenario/openscenario_utility/openscenario_utility/validation.py:68]: Isn't it too strict? Shouldn't we assure that the distance between `reach_position_conditon` and `last_destination` is lower than some kind of threshold (maybe `tolerance` from scenario)?

File: test_runner/scenario_test_runner/test/scenario/autoware-simple-test.yaml [hunk 9]
```diff
@@ -0,0 +1,167 @@
+ScenarioModifiers:
```
Comment (danielm1405) [inline test_runner/scenario_test_runner/test/scenario/autoware-simple-test.yaml:1]: Is this additional scenario necessary?

**Comment by mitsudome-r** [comment]
@WJaworskiRobotec 

Thanks for creating the PR.
Regarding this change, I've got some feedback from the author of the package (@yamacir-kit) internally, and he seems have concerns with with implementing the feature. 

Some concerns include:
* ReachPosition does not necessary have to be the final goal of the path. (e.g. There could be a new Action command implemented in the future where we dynamically give new goal during the scenario)
* As we add more success/failure conditions to the scenario, there could be AND/OR conditions, and we have to care about the design if we want to do validation of more complex conditions.

I'm sorry to tell you after all the effort you have put in already, but could you wait before you continue this task?
I'm having a meeting with him internally tomorrow to have more discussion regarding the design and give you updates.

**Comment by WJaworskiRobotec** [comment]
@mitsudome-r 

Thanks a lot for your comment. The most important part is to create base ScenarioValidator and checking selected validators in scenario_test_runner. Regarding logic implemented in example ReachPositionConditionValidator we need your feedback and suggestions. In this case, let's keep this PR as WIP, and let's agree on conditions we want to check (and maybe implement more Validators checking different conditions)

**Comment by WJaworskiRobotec** [comment]
@mitsudome-r 
As we agreed to create validator seperated from execution of the experiment, my suggestion is to keep it as a part of scenario_simulator, just launch it seperataly from running experiment. I see to possible options :
- running it as a python script with parameters : something like `python scenario_validator sceanario [SCENARIO_NAME] validators [VALIDATOR1] [VALIDATOR2] ...`
- using ros2 launch/run : something like `ros2 launch scenario_test_runner scenario_validator scenario:=... validators:= ...`

Second option might be a bit more convenient for the user, but the validator does not really start nothing ros2-based.
Do you have any opinion on this ?

File: openscenario/openscenario_utility/setup.cfg [hunk 10]
```diff
@@ -0,0 +1,4 @@
+[develop]
+script-dir=$base/lib/openscenario_utility
+[install]
+install-scripts=$base/lib/openscenario_utility
```
Comment (yamacir-kit) [inline openscenario/openscenario_utility/setup.cfg:4]: Replace the word separator hyphens in `script-dir` and `install-scripts` with underscores to suppress the warning.

```
/usr/local/lib/python3.8/dist-packages/setuptools/dist.py:697: UserWarning: Usage of dash-separated 'script-dir' will not be supported in future versions. Please use the underscore name 'script_dir' instead
```

File: openscenario/openscenario_utility/openscenario_utility/scenario_validation.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+
+class ReachPositionConditionValidator(ScenarioValidator):
+    def __init__(self):
+        super().__init__()
+        self.warning_message = "ReachPositionCondition position does not match the last position of the path"
+
+    def __call__(self, scenario):
+
+        # Search for the last Routing Action destionation:
```
Comment (yamacir-kit) [inline openscenario/openscenario_utility/openscenario_utility/scenario_validation.py:72]: Please correct the typo `destionation`.

File: openscenario/openscenario_utility/openscenario_utility/scenario_validation.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        last_destination = find_in_dict(
+            'Position', find_in_dict('RoutingAction', scenario)[-1])
+
+        # Search for ReachPositionCondition:
+        reach_position_conditon = find_in_dict(
```
Comment (yamacir-kit) [inline openscenario/openscenario_utility/openscenario_utility/scenario_validation.py:77]: Please correct the typo `conditon`.

File: openscenario/openscenario_utility/openscenario_utility/scenario_validation.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+            'Position', find_in_dict('ReachPositionCondition', scenario)[-1])
+
+        return reach_position_conditon == last_destination
```
Comment (yamacir-kit) [inline openscenario/openscenario_utility/openscenario_utility/scenario_validation.py:80]: Please correct the typo `conditon`.

File: test_runner/scenario_test_runner/test/scenario/autoware-simple-test.yaml [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+  ScenarioModifier:
+    - { name: HOGE, start: 0, step: 2, stop: 5 }
+    # - { name: FUGA, start: 1, step: 0.5, stop: 2 }
+OpenSCENARIO:
+  FileHeader:
+    revMajor: 1
+    revMinor: 0
+    date: '2021-02-01T08:31:04.600Z'
+    description: ''
+    author: ''
```
Comment (yamacir-kit) [inline test_runner/scenario_test_runner/test/scenario/autoware-simple-test.yaml:11]: What is the purpose of this newly added scenario?

**Review by yamacir-kit — CHANGES_REQUESTED** [review]

File: test_runner/scenario_test_runner/test/scenario/autoware-simple-test.yaml (hunk 14 above)
Comment (WJaworskiRobotec) [reply test_runner/scenario_test_runner/test/scenario/autoware-simple-test.yaml]: It was added for testing purpose, to validate yaml format file with reach position different from final destination. Now removed

**Review by yamacir-kit — APPROVED** [review]

**Pull request merged**